File: 202. Happy Number/isHappy.py
class Solution:
  def sumOfSquare(self, n: int) -> int:
    res = 0
    while n:
      digit = n % 10
      res += digit ** 2
      n = n // 10
    
    return res

  def isHappy(self, n: int) -> bool:
    seen = set()

    while n not in seen:
      seen.add(n)
      n = self.sumOfSquare(n)

      if n == 1:
        return True
    
    return False
  

n = 19
solution = Solution()
print("isHappy: ", solution.isHappy(n))

File: 202. Happy Number/isHappy.js
/**
 * @param {number} n
 * @return {boolean}
 */
const sumOfSquare = (n) => {
  let res = 0;

  while (n) {
    const digit = n % 10;
    res += digit ** 2;
    n = Math.floor(n / 10);
  }

  return res;
}

const isHappy = (n) => {
  const seen = new Set();

  while (!seen.has(n)) {
    seen.add(n);
    n = sumOfSquare(n);

    if (n === 1) {
      return true;
    }
  }

  return false;
};


const n = 19;
console.log("isHappy: ", isHappy(n));
